### PR TITLE
reuse `vchan::Client` to add the general header and break messages into chunks

### PR DESCRIPTION
**Pull request by ibeckermayer**
### `vchan::Client` 
An `add_header_and_chunkify` method has been added to handle adding the `Channel PDU Header` that's required for every message sent over a virtual channel, and breaking the payload up into chunks of size `<= CHANNEL_CHUNK_LEGNTH`.

### `rdpdr.rs` and `cliprdr.rs` 
`encode_message` has been replaced with `Client.add_headers_and_chunkify`, which adds the channel specific headers and then calls `vchan::Client.add_header_and_chunkify` to finalize the message.

### CODE REVIEW

File: lib/srv/desktop/rdp/rdpclient/src/vchan.rs [hunk 1]
```diff
@@ -66,6 +66,58 @@ impl Client {
...
+        let mut result = Vec::new();
+        let mut first = true;
+        while !inner.is_empty() {
+            let i = std::cmp::min(inner.len(), CHANNEL_CHUNK_LEGNTH);
+            let leftover = inner.split_off(i);
+
+            let mut channel_flags = match channel_flags {
+                Some(flags) => flags,
+                None => ChannelPDUFlags::from_bits_truncate(0),
+            };
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/vchan.rs:92]:
```suggestion
            let mut channel_flags = channel_flags.unwrap_or(ChannelPDUFlags::from_bits_truncate(0));
```

File: lib/srv/desktop/rdp/rdpclient/src/vchan.rs [hunk 2]
```diff
@@ -66,6 +66,58 @@ impl Client {
...
+
+            inner = leftover;
+        }
+
+        Ok(result)
+    }
+}
+
+/// A struct that is Encodable is able to be sent as part of an RDP PDU.
+pub trait Encodeable {
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/vchan.rs:117]: Encodeable or Encodable? You use a different spelling in the comment.

Comment (zmb3) [reply lib/srv/desktop/rdp/rdpclient/src/vchan.rs:117]: Also, can you remove this trait? I don't see it used anywhere.

File: lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs [hunk 3]
```diff
@@ -13,6 +13,7 @@
 // limitations under the License.
 
 use crate::errors::invalid_data_error;
+use crate::util::to_unicode;
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/cliprdr.rs:16]: I believe it's more idiomatic to `use crate::util` and then refer to it as `util::to_unicode`.

https://doc.rust-lang.org/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html#creating-idiomatic-use-paths

File: lib/srv/desktop/rdp/rdpclient/src/vchan.rs (hunk 2 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/vchan.rs:117]: Yeah it should be removed.

**Review by zmb3 — APPROVED** [review]

**Review by xacrimon — APPROVED** [review]

**Review by timothyb89 — APPROVED** [review]

**Pull request merged**